File: Rust/ExampleTests/common/discountinput_string.rs
#![allow(dead_code, unused_imports, unused_variables)]

#[derive(Debug, Clone, Default)]
pub struct DiscountInputString {
    pub total_price: String,
    pub discount: String,
    pub notes: String,
}

impl DiscountInputString {
    pub fn from_vec(v: &[&str]) -> Self {
        Self {
            total_price: v.get(0).copied().unwrap_or("").to_string(),
            discount: v.get(1).copied().unwrap_or("").to_string(),
            notes: v.get(2).copied().unwrap_or("").to_string(),
        }
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    pub fn from_text(text: &str) -> Self {
        let parts = crate::common::tokens::require(text, 3, "DiscountInput");
        Self {
            total_price: parts[0].clone(),
            discount: parts[1].clone(),
            notes: parts[2].clone(),
        }
    }
}

impl std::fmt::Display for DiscountInputString {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", [crate::common::tokens::token(&self.total_price), crate::common::tokens::token(&self.discount), crate::common::tokens::token(&self.notes)].join(" "))
    }
}

impl PartialEq for DiscountInputString {
    fn eq(&self, other: &Self) -> bool {
        crate::common::dnc_equal(&self.total_price, &other.total_price)
            && crate::common::dnc_equal(&self.discount, &other.discount)
            && crate::common::dnc_equal(&self.notes, &other.notes)
    }
}

File: Go/ExampleTests/common/discountinput_string.go
package common

type DiscountInputString struct {
	TotalPrice string
	Discount string
	Notes string
}

func NewDiscountInputStringFromSlice(v []string) DiscountInputString {
	s := DiscountInputString{}
	if len(v) > 0 { s.TotalPrice = v[0] }
	if len(v) > 1 { s.Discount = v[1] }
	if len(v) > 2 { s.Notes = v[2] }
	return s
}

// NewDiscountInputStringFromText builds from the text form, e.g. Money as "25 USD".
func NewDiscountInputStringFromText(text string) DiscountInputString {
	parts := RequireTokens(text, 3, "DiscountInput")
	return DiscountInputString{
		TotalPrice: parts[0],
		Discount: parts[1],
		Notes: parts[2],
	}
}

func (s DiscountInputString) String() string {
	return Token(s.TotalPrice) + " " + Token(s.Discount) + " " + Token(s.Notes)
}

func (s DiscountInputString) Equals(o DiscountInputString) bool {
	return DNCEqual(s.TotalPrice, o.TotalPrice) &&
		DNCEqual(s.Discount, o.Discount) &&
		DNCEqual(s.Notes, o.Notes)
}

func EqualDiscountInputStringSlices(a, b []DiscountInputString) bool {
	if len(a) != len(b) { return false }
	for i := range a {
		if !a[i].Equals(b[i]) { return false }
	}
	return true
}
